scenario/context.py

```diff
@@ -647,40 +611,16 @@ def run(self, event: "_Event", state: "State") -> "State":
         :arg state: the State instance to use as data source for the hook tool calls that the
             charm will invoke when handling the Event.
         """
-        if isinstance(event, _Action) or event.action:
-            raise InvalidEventError("Use run_action() to run an action event.")
-        with self._run_event(event=event, state=state) as ops:
-            ops.emit()
-        return self.output_state
-
-    def run_action(self, event: "_Event", state: "State") -> ActionOutput:
-        """Trigger a charm execution with an action event and a State.
-
-        Calling this function will call ``ops.main`` and set up the context according to the
-        specified ``State``, then emit the event on the charm.
-
-        :arg event: the action event that the charm will execute.
-        :arg state: the State instance to use as data source for the hook tool calls that the
-            charm will invoke when handling the action event.
-        """
+        if event.action:
+            self.action_history.append(Task())
```

Comment:
why are we adding there an empty task now, instead of adding a populated one, later?
Might the user access it in the meantime?

Reply:
Probably not, because it's created and 'finalised' within `run` so even if you're using the context manager you can't get to it in the test code - except I guess inside of charm code if you provide the context to the charm, which seems a very odd thing to do.

I was avoiding having to either pass a `Task` object around when it's only used for actions, or to have a `_current_task` attribute (or `_current_task_logs` and `_current_task_results` etc) that's only used for this and has no meaningful value at other times. It also felt a bit like the Juju pattern where there is a task (and operation) that evolves as it's run.

I don't love that it's like this and I'm happy to change it if there's preference for that.

Reply:
looking a bit better at the code I see that the mocking machinery needs to access the task object to put the action output, logs, etc in it, so it makes sense that it needs to be set in the context before the event processing begins. Maybe let's put a comment in to clarify, I know next time I look at it I'll be confused all over again.